### PR TITLE
support ldap:username for policy substitution

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
support ldap:username for policy substitution

## Motivation and Context
LDAPusername is the simpler form of LDAPUser (userDN),
using a simpler version is convenient from policy
conditions point of view, since these are unique id's
used for LDAP login.

## How to test this PR?
Test the new policy `{ldap:username}` as part of the policy,
NOTE: this requires changes in `mc` as well such as it needs
to compile with the latest server after this PR is merged.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#c33b528e-e3b3-49ba-a224-002844a190c8) |
<a name="c33b528e-e3b3-49ba-a224-002844a190c8"></a>
### 12390-189a969/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:31433
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp bdb20f2317ca:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... done in 2 minutes and 15 seconds
(6/15) Running healthcheck tests ... done in 5 seconds
(7/15) Running mc tests ... done in 57 seconds
(8/15) Running minio-dotnet tests ... done in 46 seconds
(9/15) Running minio-go tests ... done in 1 minutes and 54 seconds
(10/15) Running minio-java tests ... FAILED in 1 minutes and 32 seconds
{
  "name": "minio-java",
  "function": "composeObject()",
  "args": "[single source with offset]",
  "duration": 35,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = InvalidArgument, message = Range specified is not valid for source object, bucketName = minio-java-test-1qish1m, objectName = minio-java-test-hulga3, resource = /minio-java-test-1qish1m/minio-java-test-hulga3, requestId = 16831B90F3E6F0A0, hostId = 19457094-9ec9-41eb-9b62-c331e4b3ccfe)\nrequest={method=PUT, url=http://minio-dev6.minio.io:31433/minio-java-test-1qish1m/minio-java-test-hulga3?uploadId=482d0604-a1c5-4c0c-b49d-26a5f425249e&partNumber=1, headers=x-amz-copy-source: /minio-java-test-1qish1m/minio-java-test-1hlpulk\nx-amz-copy-source-range: bytes=2048-1048575\nx-amz-copy-source-if-match: cb92d17a904ccec2e6e23b8bb66245fb\nHost: minio-dev6.minio.io:31433\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210528T025640Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210528/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-copy-source-range;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=400, headers=Accept-Ranges: bytes\nContent-Length: 388\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 16831B90F3E6F0A0\nX-Xss-Protection: 1; mode=block\nDate: Fri, 28 May 2021 02:56:40 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.uploadPartCopy(MinioClient.java:4804), io.minio.MinioClient.composeObject(MinioClient.java:1431), FunctionalTest.testComposeObject(FunctionalTest.java:2120), FunctionalTest.composeObjectTests(FunctionalTest.java:2145), FunctionalTest.composeObject(FunctionalTest.java:2300), FunctionalTest.runObjectTests(FunctionalTest.java:3758), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(10/15) Running minio-js tests ... done in 46 seconds
(11/15) Running minio-py tests ... done in 2 minutes and 50 seconds
(12/15) Running s3cmd tests ... done in 17 seconds
(13/15) Running s3select tests ... done in 6 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by harshavardhana** [comment]
PTAL @Alevsk @donatello

**Review by donatello — APPROVED** [review]

File: docs/sts/rw-ldap-username.json [hunk 1]
```diff
@@ -4,10 +4,10 @@
     {
       "Effect": "Allow",
       "Action": [
-        "s3:ListBucket"
+        "s3:*"
       ],
       "Resource": [
-        "arn:aws:s3:::${ldap:user}"
+        "arn:aws:s3:::${ldap:username}/*"
```
Comment (vadmeste) [inline docs/sts/rw-ldap-username.json:10]: Oh nice, I didn't know this is possible, do we have a similar thing for regular IAM users ?

**Review by vadmeste — APPROVED** [review]
LGTM

File: docs/sts/rw-ldap-username.json (hunk 1 above)
Comment (harshavardhana) [reply docs/sts/rw-ldap-username.json:10]: yes its called `aws:username` @vadmeste and for jwt `jwt:preferred_username` ... `jwt:email`

**Pull request merged**